pyomo/gdp/plugins/bigm.py

```diff
@@ -423,77 +438,111 @@ def _bigM_relax_disjunct(self, obj, transBlock, bigM, suffix_list):
         # comparing the two relaxations.
         #
         # Transform each component within this disjunct
-        self._transform_block_components(obj, obj, infodict, bigM, suffix_list)
+        self._transform_block_components(obj, obj, bigM, arg_list, suffix_list)
 
-        # deactivate disjunct so we know we've relaxed it
+        # deactivate disjunct to keep the writers happy
         obj._deactivate_without_fixing_indicator()
-        infodict['relaxed'] = True
 
-    def _transform_block_components(self, block, disjunct, infodict,
-                                    bigM, suffix_list):
-        # Look through the component map of block and transform
+    def _transform_block_components(self, block, disjunct, bigM, arg_list,
+                                    suffix_list):
+        # We first need to find any transformed disjunctions that might be here
+        # because we need to move their transformation blocks up onto the parent
+        # block before we transform anything else on this block 
+        destinationBlock = disjunct._transformation_block().parent_block()
+        for obj in block.component_data_objects(
+                Disjunction, 
+                sort=SortComponents.deterministic, 
+                descend_into=(Block)):
+            if obj.algebraic_constraint is None:
+                # This could be bad if it's active since that means its
+                # untransformed, but we'll wait to yell until the next loop
+                continue
+            # get this disjunction's relaxation block.
+            transBlock = obj.algebraic_constraint().parent_block()
+            
+            # move transBlock up to parent component
+            self._transfer_transBlock_data(transBlock, destinationBlock)
+            # we leave the transformation block because it still has the XOR
+            # constraints, which we want to be on the parent disjunct.
+
+        # Now look through the component map of block and transform
         # everything we have a handler for. Yell if we don't know how
         # to handle it.
         for name, obj in list(iteritems(block.component_map())):
-            if hasattr(obj, 'active') and not obj.active:
+            # This means non-ActiveComponent types cannot have handlers
```

Comment:
Should we just do:
```python
for obj in block.component_objects(active=True, descend_into=False): 
```
instead?  Since we will not complain about non-ActiveComponent objects, we might as well just not visit them.

Also, since we are skipping non-active components, we can remove the handler for Var, Expression, and Suffix, right?

Reply:
I agree about rewriting the loop, but we still hit Var, Expression and Suffix in what you suggest. So we would still need the handlers. The only way I see to get rid of the handlers is if we only iterate through stuff we need to transform. But then we would fail silently if there was something odd, so I'm tempted to leave it.